collect_app/src/main/java/org/odk/collect/android/formentry/backgroundlocation/BackgroundLocationManager.java

```diff
@@ -0,0 +1,284 @@
+            currentState = BackgroundLocationState.STOPPED;
+        }
+    }
+
+    public void backgroundLocationPreferenceToggled() {
+        switch (currentState) {
+            case RECEIVING_LOCATIONS:
+                if (!helper.isBackgroundLocationPreferenceEnabled()) {
+                    helper.logAuditEvent(AuditEvent.AuditEventType.LOCATION_TRACKING_DISABLED);
+                    stopLocationRequests();
+                }
+                break;
+
+            case STOPPED:
+                if (helper.isBackgroundLocationPreferenceEnabled()) {
+                    helper.logAuditEvent(AuditEvent.AuditEventType.LOCATION_TRACKING_ENABLED);
+                    startLocationRequests();
+                }
+                break;
+        }
+    }
+
+    public void locationProvidersChanged() {
+        switch (currentState) {
+            case RECEIVING_LOCATIONS:
+            case STOPPED:
+                if (locationClient.isLocationAvailable()) {
+                    helper.logAuditEvent(AuditEvent.AuditEventType.LOCATION_PROVIDERS_ENABLED);
+                } else {
+                    helper.logAuditEvent(AuditEvent.AuditEventType.LOCATION_PROVIDERS_DISABLED);
+                }
+                break;
+        }
+    }
+
+    private void startLocationRequests() {
+        if (helper.currentFormAuditsLocation()
+                && helper.isBackgroundLocationPreferenceEnabled()
+                && helper.arePlayServicesAvailable()
+                && helper.isAndroidLocationPermissionGranted()) {
+            AuditConfig auditConfig = helper.getCurrentFormAuditConfig();
+
+            locationClient.setListener(this);
+            locationClient.setPriority(auditConfig.getLocationPriority());
+            locationClient.setUpdateIntervals(auditConfig.getLocationMinInterval(), auditConfig.getLocationMinInterval());
+            locationClient.start();
+
+            currentState = BackgroundLocationState.RECEIVING_LOCATIONS;
+        }
+    }
+
+    private void stopLocationRequests() {
+        locationClient.setListener(null);
+        locationClient.stop();
+
+        currentState = BackgroundLocationState.STOPPED;
+    }
+
+    @Override
+    public void onLocationChanged(Location location) {
+        if (currentState == BackgroundLocationState.RECEIVING_LOCATIONS) {
+            helper.provideLocationToAuditLogger(location);
+        }
+    }
+
+    @Override
+    public void onClientStart() {
+        locationClient.requestLocationUpdates(locationListener);
+    }
+
+    @Override
+    public void onClientStartFailure() {
+
+    }
+
+    @Override
+    public void onClientStop() {
+
+    }
+
+    private enum BackgroundLocationState {
+        /** The current form does not track background location (also the case if the current form
+         * is not set yet */
+        NOT_APPLICABLE,
+
+        /** The current form tracks background location and a message hasn't been shown to the user **/
+        APPLICABLE,
```

Comment:
The naming of `APPLICABLE` and `NOT_APPLICABLE` took me a little while to get my head around. Maybe that's just me but could it be `FORM_NEEDS_LOCATION` or something?

Reply:
How about `NO_BACKGROUND_LOCATION_NEEDED` and `PENDING_PRECONDITION_CHECKS`?